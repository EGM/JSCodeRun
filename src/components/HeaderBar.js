
import React, { Component } from 'react';
import { 
    StyleSheet, 
    Text, 
    View, 
    TouchableOpacity, 
    TouchableHighlight, 
    Modal, 
    FlatList,
    Button,
    Image
} from 'react-native';
import codeExamples from '../constants/examples.js';

const styles = StyleSheet.create({
    container: {
        height: 80,
        width: '100%',
        backgroundColor: '#ddd'
    },
    headerBar: {
        flex: 1,
        flexDirection: 'row',
        alignItems: 'flex-end',
        justifyContent: 'space-between',
        marginHorizontal: 5
    },
    headerBarItem: {
        padding: 5
    },
    listItem: {
        paddingHorizontal: 15,
        paddingVertical: 7
    },
    separator: {
        height: 1,
        width: "94%",
        backgroundColor: "#CED0CE",
        marginLeft: "3%"
    },
    modalContainer: {
        marginTop: 20
    },
    modalClose: {
        borderBottomWidth: 1,
        borderColor:'#d6d7da'
    },
    listItemSmallText: {
        marginLeft: 10,
        opacity: 0.7
    },
    runButton: {
        height: 40,
        width: 40
    }
});

const ListItem = ({item, onPress}) =>
    <TouchableHighlight style={styles.listItem}
        activeOpacity={0.2}
        underlayColor="#eee"
        onPress={() => onPress(item.code)}>
        <View>
            <Text>{item.key}</Text>
            <Text style={styles.listItemSmallText}>Runtime:{item.runtimeComplexity}, Space:{item.spaceComplexity}</Text>
        </View>
    </TouchableHighlight>

const Separator = () => <View style={styles.separator}/>

export default class HeaderBar extends Component {
  constructor(props) {
      super(props);
      this.state = {
          isExamplesModalDisplayed: false
      };
  }

  toggleExamplesModal = () => {
    this.setState({
        isExamplesModalDisplayed: !this.state.isExamplesModalDisplayed
    })
  }

  render() {
    return <View style={styles.container}>
        <Modal
            animationType="slide"
            transparent={false}
            visible={this.state.isExamplesModalDisplayed}
            onRequestClose={() => {this.toggleExamplesModal()}}
        >
            <View style={styles.modalContainer}>
                
                <Button style={styles.modalClose} onPress={() => this.toggleExamplesModal()} title="Close âœ–" />
                    <FlatList
                        data={codeExamples}
                        ItemSeparatorComponent={() => <Separator />}
                        ListHeaderComponent={() => <Separator />}
                        ListFooterComponent={() => <Separator />}
                        keyExtractor={item => item.key}
                        renderItem={({item}) => <ListItem item={item} onPress={code => {this.props.generateCode(code); this.toggleExamplesModal()}}/>}
                    />
            </View>
        </Modal>
        <View style={styles.headerBar}>
            <TouchableOpacity style={styles.headerBarItem} onPress={() => {this.toggleExamplesModal()}}>
                <Text>Code examples</Text>
            </TouchableOpacity>
            <TouchableOpacity style={styles.headerBarItem} onPress={() => {this.props.runCode()}}>
                <Image style={styles.runButton} source={require('../../assets/run.png')} />
            </TouchableOpacity>

        </View>
    </View>
  }
}