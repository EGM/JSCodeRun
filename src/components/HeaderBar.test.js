import React, { Component } from 'react';
import renderer from 'react-test-renderer';
import HeaderBar from './HeaderBar';
import exampleData from '../constants/examples';

jest.mock('../constants/examples', () => [
  {
    key: 'Example 1',
    runtimeComplexity: 'O(1)',
    spaceComplexity: 'O(n)',
    code: 'someCode();'
  }, {
    key: 'Example 2',
    code: 'someMoreCode();'
  },
])

describe('HeaderBar', () => {
  it('should render the Header Bar correctly', () => {
    const rendered = renderer.create(
      <HeaderBar 
        runCode={() => {}}
        generateCode={() => {}}
      />
    ).toJSON();
    expect(rendered).toMatchSnapshot();
  });
});