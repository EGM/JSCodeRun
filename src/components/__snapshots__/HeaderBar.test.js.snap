// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderBar should render the Header Bar correctly 1`] = `
<View
  style={
    Object {
      "height": 55,
      "marginTop": 20,
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginHorizontal": 10,
      }
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "flex": 1,
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Image
        source={
          Object {
            "testUri": "../../../assets/hamburger.png",
          }
        }
        style={
          Object {
            "height": 35,
            "tintColor": "white",
            "width": 35,
          }
        }
      />
    </View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "white",
          "flexGrow": 1,
          "fontSize": 22,
          "textAlign": "center",
        }
      }
    >
      JS Code Run
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "flex-end",
        }
      }
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
            "padding": 5,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Image
          source={
            Object {
              "testUri": "../../../assets/save-icon.png",
            }
          }
          style={
            Object {
              "height": 25,
              "tintColor": "white",
              "width": 25,
            }
          }
        />
      </View>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Image
          source={
            Object {
              "testUri": "../../../assets/run.png",
            }
          }
          style={
            Object {
              "height": 35,
              "tintColor": "white",
              "width": 35,
            }
          }
        />
      </View>
    </View>
  </View>
</View>
`;
